Cardano improvements

refi93 opened the pull request:
- unified address export screen with other cryptourrencies (displaying QR/string)
- changed UI of signing transactions to be more similar to BTC
- better grouping of information when signing Tx and dropping info about network and tx size which we deem redundant

refi93 commented:
We thought a bit more about the "network" attribute being sent with the transaction and we suggest changing it directly to protocol magic - the integer being serialized in the signature, specific for the network. This allows trezor to sign transactions for custom networks. There are already PRs on trezor-common and trezor-connect repos to update the name of the field in the calls: https://github.com/trezor/trezor-common/pull/244 , https://github.com/trezor/connect/pull/272

To prevent an attack where the adversary tricks the signer into taking transaction inputs from the mainnet when sending a testnet transaction we suggest displaying the network explicitly. It's probably a hypothetical attack, but better stay on the safe side.

ppershing commented on src/apps/cardano/sign_tx.py:
```diff
@@ -118,39 +110,29 @@
-        else:
-            raise wire.ProcessError("Unknown network index %d" % network)
+
+        self.network_name = KNOWN_PROTOCOL_MAGICS.get(protocol_magic) or 'Unknown'
```
Unless micrypython differs from python here, we should be able specify default value in `get` like this:

```
>>> {'xx': 1}.get('xx', 'def')
1
>>> {'xx': 1}.get('yy', 'def')
'def'
```

prusnak replied:
This works in uPy.

refi93 replied:
fixed

ppershing commented on src/apps/cardano/sign_tx.py:
```diff
@@ -7,69 +7,61 @@
     outcoins: list,
-    change_derivation_paths: list,
-    change_coins: list,
     fee: float,
```
Hmm, why is fee `float` here?

refi93 replied:
good point, it's an int actually, fixed

jpochyla commented on src/apps/cardano/layout/__init__.py:
```diff
@@ -3,11 +3,67 @@
+
+    paginator = paginate(create_renderer(ConfirmDialog), len(pages), const(0), pages)
+    return await ctx.wait(paginator) == CONFIRMED
+    return True
```
Dead code.

refi93 commented:
fixed code style and removed dead code. I did not manage to run the tests in python-trezor though: https://github.com/trezor/python-trezor/pull/342 seems like emulation of swipe and confirm does not work for test_msg_cardano_sign_transaction.py. It waits for manually swiping and confirming. And I did not manage to understand from the code, how to properly emulate those events. What am I missing?

jpochyla commented:
@refi93 We are testing the swipable dialogs this way: https://github.com/trezor/python-trezor/blob/master/trezorlib/tests/device_tests/test_msg_resetdevice_t2.py#L56, it should work in all cases that use `trezor.ui.scroll`.

prusnak commented:
FYI I just merged https://github.com/trezor/trezor-common/pull/244 into master

refi93 commented:
I added ButtonRequests to `confirm_transaction` and `confirm_sendind` screens so it is testable, hence the force push.

refi93 commented:
updated trezor-common submodule

refi93 commented:
I recently added output address validation, to solve an issue we recently came across - i.e. that a specially constructed invalid address may cause the raw transaction to have more outputs than advertised by trezor ui.

refi93 commented:
I recently added output address validation
